File: ios/Sophia/Services/SocialService.swift
import Foundation
import Observation
import Supabase

/// Période du classement d'amis.
nonisolated enum FriendsLeaderboardPeriod: String, CaseIterable, Sendable {
    case week
    case all

    var rpcValue: String {
        switch self {
        case .week: "week"
        case .all: "all"
        }
    }
}

nonisolated struct FriendLeaderboardEntry: Identifiable, Decodable, Sendable, Equatable {
    let userId: UUID
    let handle: String
    let xp: Int
    let isMe: Bool

    var id: UUID { userId }

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case handle
        case xp
        case isMe = "is_me"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        userId = try c.decode(UUID.self, forKey: .userId)
        handle = try c.decode(String.self, forKey: .handle)
        if let intXP = try? c.decode(Int.self, forKey: .xp) {
            xp = intXP
        } else if let longXP = try? c.decode(Int64.self, forKey: .xp) {
            xp = Int(longXP)
        } else {
            xp = 0
        }
        isMe = try c.decode(Bool.self, forKey: .isMe)
    }
}

nonisolated struct FriendPublicStats: Decodable, Sendable, Equatable {
    let userId: UUID
    let handle: String
    let globalXP: Int
    let streak: Int
    let coursesCompleted: Int
    let quizzesCompleted: Int

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case handle
        case globalXP = "global_xp"
        case streak
        case coursesCompleted = "courses_completed"
        case quizzesCompleted = "quizzes_completed"
    }
}

nonisolated enum SocialError: Error, Sendable, Equatable {
    case notSignedIn
    case invalidHandle
    case handleTaken
    case userNotFound
    case cannotAddSelf
    case notFriends
    case underlying(String)

    static func from(_ error: Error) -> SocialError {
        let message = error.localizedDescription.lowercased()
        if message.contains("invalid_handle") { return .invalidHandle }
        if message.contains("handle_taken") { return .handleTaken }
        if message.contains("user_not_found") { return .userNotFound }
        if message.contains("cannot_add_self") { return .cannotAddSelf }
        if message.contains("not_friends") { return .notFriends }
        if message.contains("not_authenticated") { return .notSignedIn }
        return .underlying(error.localizedDescription)
    }
}

/// Handles (@), amis et classement XP via Supabase.
@Observable
@MainActor
final class SocialService {
    static let shared = SocialService()

    private(set) var myHandle: String?
    private(set) var leaderboard: [FriendLeaderboardEntry] = []
    private(set) var period: FriendsLeaderboardPeriod = .week
    private(set) var isLoadingHandle = false
    private(set) var isLoadingLeaderboard = false

    private let client = SupabaseManager.shared

    private init() {}

    // MARK: - Profile handle

    func refreshMyHandle() async {
        guard AuthService.shared.isSignedIn,
              let userID = AuthService.shared.currentUser?.id.uuidString else {
            myHandle = nil
            return
        }
        isLoadingHandle = true
        defer { isLoadingHandle = false }
        do {
            struct Row: Decodable { let handle: String }
            let rows: [Row] = try await client
                .from("profiles")
                .select("handle")
                .eq("id", value: userID)
                .limit(1)
                .execute()
                .value
            myHandle = rows.first?.handle
        } catch {
            #if DEBUG
            print("[Social] refreshMyHandle failed: \(error)")
            #endif
        }
    }

    @discardableResult
    func updateHandle(_ raw: String) async throws -> String {
        guard AuthService.shared.isSignedIn else { throw SocialError.notSignedIn }
        struct Params: Encodable { let new_handle: String }
        do {
            let updated: String = try await client
                .rpc("update_my_handle", params: Params(new_handle: raw))
                .execute()
                .value
            myHandle = updated
            return updated
        } catch {
            throw SocialError.from(error)
        }
    }

    // MARK: - Friends

    func addFriend(handle raw: String) async throws {
        guard AuthService.shared.isSignedIn else { throw SocialError.notSignedIn }
        struct Params: Encodable { let target_handle: String }
        do {
            let _: UUID = try await client
                .rpc("add_friend_by_handle", params: Params(target_handle: raw))
                .execute()
                .value
            await refreshLeaderboard()
        } catch {
            throw SocialError.from(error)
        }
    }

    func removeFriend(userId: UUID) async throws {
        guard AuthService.shared.isSignedIn else { throw SocialError.notSignedIn }
        struct Params: Encodable { let friend: UUID }
        do {
            try await client
                .rpc("remove_friend", params: Params(friend: userId))
                .execute()
            await refreshLeaderboard()
        } catch {
            throw SocialError.from(error)
        }
    }

    func setPeriod(_ newPeriod: FriendsLeaderboardPeriod) async {
        guard period != newPeriod else { return }
        period = newPeriod
        await refreshLeaderboard()
    }

    func refreshLeaderboard() async {
        guard AuthService.shared.isSignedIn else {
            leaderboard = []
            return
        }
        isLoadingLeaderboard = true
        defer { isLoadingLeaderboard = false }
        struct Params: Encodable { let period: String }
        do {
            let rows: [FriendLeaderboardEntry] = try await client
                .rpc("friends_leaderboard", params: Params(period: period.rpcValue))
                .execute()
                .value
            leaderboard = rows
        } catch {
            #if DEBUG
            print("[Social] refreshLeaderboard failed: \(error)")
            #endif
        }
    }

    func friendStats(userId: UUID) async throws -> FriendPublicStats {
        guard AuthService.shared.isSignedIn else { throw SocialError.notSignedIn }
        struct Params: Encodable { let target: UUID }
        do {
            let rows: [FriendPublicStats] = try await client
                .rpc("friend_public_stats", params: Params(target: userId))
                .execute()
                .value
            guard let stats = rows.first else { throw SocialError.userNotFound }
            return stats
        } catch let social as SocialError {
            throw social
        } catch {
            throw SocialError.from(error)
        }
    }

    // MARK: - XP events

    /// Enregistre un gain d'XP global pour le classement hebdomadaire (best-effort).
    func logXPEvent(amount: Int, reason: String?) {
        guard amount > 0, AuthService.shared.isSignedIn else { return }
        Task { [weak self] in
            await self?.pushXPEvent(amount: amount, reason: reason)
        }
    }

    private func pushXPEvent(amount: Int, reason: String?) async {
        struct Params: Encodable {
            let amount: Int
            let reason: String?
        }
        do {
            try await client
                .rpc("log_xp_event", params: Params(amount: amount, reason: reason))
                .execute()
        } catch {
            #if DEBUG
            print("[Social] logXPEvent failed: \(error)")
            #endif
        }
    }

    /// Charge handle + classement après connexion / au focus profil.
    func refreshAll() async {
        await refreshMyHandle()
        await refreshLeaderboard()
    }

    func clearLocalState() {
        myHandle = nil
        leaderboard = []
    }
}

File: android/app/src/main/java/app/rork/sophia/data/SocialService.kt
package app.rork.sophia.data

import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

@Serializable
data class FriendLeaderboardEntry(
    val user_id: String,
    val handle: String,
    val xp: Int = 0,
    val is_me: Boolean = false,
)

@Serializable
data class FriendRequest(
    val request_id: String,
    val user_id: String,
    val handle: String,
)

@Serializable
data class FriendPublicStats(
    val user_id: String,
    val handle: String,
    val global_xp: Int = 0,
    val streak: Int = 0,
    val courses_completed: Int = 0,
    val quizzes_completed: Int = 0,
)

enum class FriendsLeaderboardPeriod(val rpcValue: String) {
    WEEK("week"),
    ALL("all"),
}

class SocialService(private val auth: AuthService) {
    private val _myHandle = MutableStateFlow<String?>(null)
    val myHandle: StateFlow<String?> = _myHandle.asStateFlow()

    private val _pending = MutableStateFlow<List<FriendRequest>>(emptyList())
    val pendingRequests: StateFlow<List<FriendRequest>> = _pending.asStateFlow()

    private val _leaderboard = MutableStateFlow<List<FriendLeaderboardEntry>>(emptyList())
    val leaderboard: StateFlow<List<FriendLeaderboardEntry>> = _leaderboard.asStateFlow()

    private val _period = MutableStateFlow(FriendsLeaderboardPeriod.WEEK)
    val period: StateFlow<FriendsLeaderboardPeriod> = _period.asStateFlow()

    /** StringStore key, so the message follows the app language instead of being baked in French. */
    private val _errorKey = MutableStateFlow<String?>(null)
    val errorKey: StateFlow<String?> = _errorKey.asStateFlow()

    private val db get() = SupabaseManager.client.postgrest

    fun setPeriod(period: FriendsLeaderboardPeriod) {
        _period.value = period
    }

    fun clearError() {
        _errorKey.value = null
    }

    fun clearLocalState() {
        _myHandle.value = null
        _pending.value = emptyList()
        _leaderboard.value = emptyList()
        _errorKey.value = null
    }

    suspend fun refreshAll() {
        if (!auth.isSignedIn) return
        refreshMyHandle()
        refreshPendingRequests()
        refreshLeaderboard()
    }

    suspend fun refreshMyHandle() {
        val userId = auth.userId.value ?: return
        runCatching {
            @Serializable
            data class Row(val handle: String? = null)
            val row = SupabaseManager.client.from("profiles")
                .select { filter { eq("id", userId) } }
                .decodeSingleOrNull<Row>()
            _myHandle.value = row?.handle
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun updateHandle(raw: String): Result<String> {
        @Serializable
        data class Params(val new_handle: String)
        val handle = sanitizeHandle(raw)
        _errorKey.value = null
        return runCatching {
            val updated = db.rpc("update_my_handle", Params(new_handle = handle)).decodeAs<String>()
            _myHandle.value = updated
            updated
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    /** @return [SendRequestOutcome] so the UI can tell "request sent" from "you are now friends". */
    suspend fun sendFriendRequest(rawHandle: String): Result<SendRequestOutcome> {
        @Serializable
        data class Params(val target_handle: String)
        val handle = sanitizeHandle(rawHandle)
        _errorKey.value = null
        if (handle.isEmpty()) {
            _errorKey.value = "friends.error.invalidHandle"
            return Result.failure(IllegalArgumentException("invalid_handle"))
        }
        return runCatching {
            val outcome = db.rpc("send_friend_request", Params(target_handle = handle)).decodeAs<String>()
            refreshPendingRequests()
            refreshLeaderboard()
            if (outcome == "accepted") SendRequestOutcome.AUTO_ACCEPTED else SendRequestOutcome.SENT
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun respondToRequest(requestId: String, accept: Boolean): Result<Unit> {
        @Serializable
        data class Params(val request_id: String, val accept: Boolean)
        _errorKey.value = null
        return runCatching {
            db.rpc("respond_friend_request", Params(request_id = requestId, accept = accept))
            refreshPendingRequests()
            refreshLeaderboard()
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun refreshPendingRequests() {
        runCatching {
            _pending.value = db.rpc("pending_friend_requests").decodeList()
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun removeFriend(userId: String): Result<Unit> {
        @Serializable
        data class Params(val friend: String)
        _errorKey.value = null
        return runCatching {
            db.rpc("remove_friend", Params(friend = userId))
            refreshLeaderboard()
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun refreshLeaderboard() {
        @Serializable
        data class Params(val period: String)
        val period = _period.value.rpcValue
        runCatching {
            _leaderboard.value = db.rpc("friends_leaderboard", Params(period = period)).decodeList()
        }.onFailure { _errorKey.value = errorKeyFor(it.message) }
    }

    suspend fun friendStats(userId: String): FriendPublicStats? {
        @Serializable
        data class Params(val target: String)
        return runCatching {
            db.rpc("friend_public_stats", Params(target = userId)).decodeList<FriendPublicStats>().firstOrNull()
        }.getOrNull()
    }

    suspend fun logXPEvent(amount: Int, reason: String?) {
        if (!auth.isSignedIn || amount <= 0) return
        @Serializable
        data class Params(val amount: Int, val reason: String? = null)
        runCatching {
            db.rpc("log_xp_event", Params(amount = amount, reason = reason))
        }
    }

    companion object {
        fun sanitizeHandle(raw: String): String =
            raw.trim().removePrefix("@").lowercase()

        /** Postgres raises bare codes; turn them into StringStore keys. */
        fun errorKeyFor(message: String?): String {
            val m = message.orEmpty()
            return when {
                "invalid_handle" in m -> "friends.error.invalidHandle"
                "handle_taken" in m -> "friends.error.handleTaken"
                "user_not_found" in m -> "friends.error.userNotFound"
                "cannot_add_self" in m -> "friends.error.cannotAddSelf"
                "already_friends" in m -> "friends.error.alreadyFriends"
                "request_already_sent" in m -> "friends.error.requestAlreadySent"
                "request_not_found" in m -> "friends.error.requestNotFound"
                "not_friends" in m -> "friends.error.notFriends"
                "not_authenticated" in m -> "friends.error.notSignedIn"
                else -> "friends.error.generic"
            }
        }
    }
}

enum class SendRequestOutcome { SENT, AUTO_ACCEPTED }
